### PR TITLE
[SPARK-13776][WebUI]Limit the max number of acceptors and selectors for Jetty

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

As each acceptor/selector in Jetty will use one thread, the number of threads should at least be the number of acceptors and selectors plus 1. Otherwise, the thread pool of Jetty server may be exhausted by acceptors/selectors and not be able to response any request.

To avoid wasting threads, the PR limits the max number of acceptors and selectors and also updates the max thread number if necessary.
## How was this patch tested?

Just make sure we don't break any existing tests

### CODE REVIEW

**Comment by zsxwing** [comment]
Another option is just hard-core the number of acceptors to a small number. I don't think there will be a lot of connections to access the Web UI at the same time.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 1]
```diff
@@ -270,9 +270,21 @@ private[spark] object JettyUtils extends Logging {
...
       gzipHandlers.foreach(collection.addHandler)
       connectors.foreach(_.setHost(hostName))
+      // As each Acceptor will use one thread, the number of threads should at least be the number
+      // of acceptors plus 1.
+      val minThreads =
+        connectors.collect { case c: AbstractConnector => c }.map(_.getAcceptors).sum + 1
       server.setConnectors(connectors.toArray)
 
       val pool = new QueuedThreadPool
+      var maxThreads = conf.getInt("spark.ui.threads", pool.getMaxThreads)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Do we even need config here? when would I otherwise want to control this? It seems like you want to always make the pool have at least acceptors+1 threads.

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: I was worried that 1 thread for processing requests may be not enough.

@srowen what do you think about `just hard-core the max number of acceptors to a small number`? Maybe just 8?

The default configuration of Jetty seems good to a web application that supports massive  connections. However, Spark Web UI doesn't need to support so many users, so it's a waste to create many threads.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Agree, if it's just for web UIs then I see no need for a large number of acceptors. I think your change is still needed to make sure the pool has enough threads in any event? though maybe the config is unnecessary.

**Comment by zsxwing** [comment]
Updated to use a small acceptor number

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 2]
```diff
@@ -270,9 +270,19 @@ private[spark] object JettyUtils extends Logging {
 
       gzipHandlers.foreach(collection.addHandler)
       connectors.foreach(_.setHost(hostName))
+      // As each Acceptor will use one thread, the number of threads should at least be the number
+      // of acceptors plus 1. (See SPARK-13776)
+      var minThreads = 1
+      connectors.collect { case c: AbstractConnector => c }.foreach { c =>
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Nit: the Scala math class is `math` vs `Math` but I wouldn't change it just for that. I think this looks like the right kind of approach.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 3]
```diff
@@ -270,9 +270,25 @@ private[spark] object JettyUtils extends Logging {
...
       connectors.foreach(_.setHost(hostName))
+      // As each acceptor and each selector will use one thread, the number of threads should at
+      // least be the number of acceptors and selectors plus 1. (See SPARK-13776)
+      var minThreads = 1
+      connectors.foreach { c =>
+        // Currently we only use "SelectChannelConnector"
+        val connector = c.asInstanceOf[SelectChannelConnector]
+        // Limit the max acceptor number to 8 so that we don't waste a lot of threads
+        connector.setAcceptors(math.min(connector.getAcceptors, 8))
+        // The number of selectors always equals to the number of acceptors
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: See https://github.com/eclipse/jetty.project/blob/jetty-8.1.14.v20131031/jetty-server/src/main/java/org/eclipse/jetty/server/nio/SelectChannelConnector.java#L260

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 4]
```diff
@@ -270,9 +270,25 @@ private[spark] object JettyUtils extends Logging {
 
       gzipHandlers.foreach(collection.addHandler)
       connectors.foreach(_.setHost(hostName))
+      // As each acceptor and each selector will use one thread, the number of threads should at
+      // least be the number of acceptors and selectors plus 1. (See SPARK-13776)
+      var minThreads = 1
+      connectors.foreach { c =>
```
Comment (jaceklaskowski) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Use `collect` or "cast" using pattern matching on type.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 5]
```diff
@@ -270,9 +270,25 @@ private[spark] object JettyUtils extends Logging {
...
+        minThreads += connector.getAcceptors * 2
+      }
       server.setConnectors(connectors.toArray)
 
       val pool = new QueuedThreadPool
+      if (serverName.nonEmpty) {
+        pool.setName(serverName)
+      }
+      pool.setMaxThreads(math.max(pool.getMaxThreads, minThreads))
+      pool.setMinThreads(math.min(pool.getMinThreads, pool.getMaxThreads))
```
Comment (jaceklaskowski) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: When could `getMinThreads` be greater than `getMaxThreads`? Why don't you use `minThreads` here, too?

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 4 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: I want to crash the app if we add a non-SelectChannelConnector in future. It's better than being silent.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 5 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: The default `pool.getMinThreads` may be greater than `minThreads` here. So I use `pool.getMaxThreads` here.

Comment (jaceklaskowski) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: So, could getMinThreads be greater than getMaxThreads ever? Why are you doing the line?

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: If `minThreads=4, getMaxThreads=100, getMinThreads=8`, I want to call `setMaxThreads(100), setMinThreads(8)`.
If `minThreads=200, getMaxThreads=100, getMinThreads=8`, I want to call `setMaxThreads(200), setMinThreads(8)`.
If `minThreads=10, getMaxThreads=100, getMinThreads=8`, I want to call `setMaxThreads(200), setMinThreads(8)`.

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Never mind. I see. I don't need to `setMinThreads`.

**Comment by zsxwing** [comment]
@srowen what do you think now?

File: core/src/main/scala/org/apache/spark/ui/WebUI.scala [hunk 6]
```diff
@@ -134,7 +134,7 @@ private[spark] abstract class WebUI(
   def bind() {
     assert(!serverInfo.isDefined, "Attempted to bind %s more than once!".format(className))
     try {
-      var host = Option(conf.getenv("SPARK_LOCAL_IP")).getOrElse("0.0.0.0")
+      val host = Option(conf.getenv("SPARK_LOCAL_IP")).getOrElse("0.0.0.0")
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/WebUI.scala]: OK but was this change intentional? rest seems OK

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/WebUI.scala]: Oh, never mind. I will remove it.

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by srowen** [comment]
LGTM, and the original reporter confirmed it works. I'm merging to master. I'm open to an argument that this should go into 1.6 though the scope of the problem is quite limited, so maybe isn't worth the subtle behavior change.

**Comment by zsxwing** [comment]
Agree that this doesn't need to go into 1.6.
